File: opensource/kernel/src/main/java/com/sos/scheduler/engine/eventbus/HotEventBus.java
package com.sos.scheduler.engine.eventbus;

import java.util.Collection;

import javax.annotation.Nullable;

import org.apache.log4j.Logger;

import com.sos.scheduler.engine.kernel.event.EventSubsystem;

public class HotEventBus extends AbstractEventBus {
    private static final Logger logger = Logger.getLogger(HotEventBus.class);

    @Nullable private Event currentEvent = null;

    public HotEventBus() {
        super(HotEventHandler.class);
    }

    @Override public final void publish(Event e) {
        publish(e, calls(e));
    }

    final void publish(Event e, Collection<Call> calls) {
        if (currentEvent != null)
            handleRecursiveEvent(e);
        else
            dispatchNonrecursiveEvent(e, calls);
    }

    //@Override public final void dispatchEvents() {}

    private void handleRecursiveEvent(Event e) {
        try {
            // Kein log().error(), sonst gibt es wieder eine Rekursion
            throw new Exception(EventSubsystem.class.getSimpleName() + ".publish("+e+"): ignoring the event triggered by handling the event '"+currentEvent+"'");
        }
        catch (Exception x) {
            logger.error(x, x);
        }
    }

    private void dispatchNonrecursiveEvent(Event e, Collection<Call> calls) {
        currentEvent = e;
        try {
            for (Call c: calls)
                dispatchCall(c);
        }
        finally {
            currentEvent = null;
        }
    }
}

File: engine-kernel/src/main/scala/com/sos/scheduler/engine/eventbus/HotEventBus.scala
package com.sos.scheduler.engine.eventbus

import com.sos.scheduler.engine.common.scalautil.Logger
import com.sos.scheduler.engine.data.event.{Event, KeyedEvent}
import com.sos.scheduler.engine.eventbus.HotEventBus.logger
import com.sos.scheduler.engine.eventbus.annotated.HotMethodEventSubscriptionFactory
import java.util.Objects.requireNonNull
import java.util.concurrent.atomic.AtomicReference

final class HotEventBus extends AbstractEventBus {

  protected def factory = HotMethodEventSubscriptionFactory.singleton

  private val currentEvent = new AtomicReference[KeyedEvent[Event]](null)  // This code is not thread-safe! Just in case someone tries.

  def publish(e: KeyedEvent[Event]) =
    if (currentEvent.compareAndSet(null, requireNonNull(e)))
      try for (c ← calls(e)) dispatchCall(c)
      finally currentEvent.set(null)
    else
      logger.error("Ignored", new RuntimeException(s"HotEventBus.publish($e): ignoring this event while handling the event '${currentEvent.get}'"))
}

object HotEventBus {
  private val logger = Logger(getClass)
}
